### DISCUSSION TITLE
Hide an interaction/specific residue inclusively?

### DISCUSSION THREAD

**Discussion by ONISniperOne** [chemosim-lab/ProLIF#176] [Q&A] [answered]
I found this as part of the protein-ligand tutorial but it's for the dataframe

```
# hide an interaction type (Hydrophobic)
df.drop("Hydrophobic", level="interaction", axis=1).head(5)
```

Is there a way to not include an interaction that's part of 2 separate categories such as hide all VDW interactions between ligand and residue 209? Still show other types of interaction with 209 and other VDW interactions with other residues.

I was just wondering if there's a way to implement this in pandas and the 2D/3D lignetwork.

**Comment by cbouy** [comment, marked as answer]
For the dataframe only you can pass a list of tuples of columns to drop e.g.
```python
df.drop([("LIG1", "ALA209", "VdWContact")], axis=1)
```
So you need to specify the 3 levels (ligand, protein residue, and interaction type) for each residue and interaction that you want to hide.

For the lignetwork, there's no direct way of doing that, but you can directly play with the `fp.ifp` object which stores the data (make sure to save the fingerprint as a pickle file first):
```python
# save
fp.to_pickle("fingerprint.pkl")

# setup things to clean
to_cleanup = [
    ("LIG1.G", "TYR38.A", "Hydrophobic"),
    ("LIG1.G", "TYR38.A", "VdWContact"),
]

# start cleanup
df = fp.to_dataframe()
for lig, residue, interaction in to_cleanup:
    # locate all frames with this combination
    frames = df.index[df.loc[:, (lig, residue, interaction)]]
    for frame in frames:
        # delete interaction from registry
        del fp.ifp[frame][(lig, residue)][interaction]
```
It will affect both `to_dataframe` and `plot_lignetwork`

**Reply by ONISniperOne** [reply]
I tried this one first, keeping the original fingerprint, and paired it with groupby
```
# Drop the specified column first
df_filtered = df.drop([("UNL210", "SIA209", "VdWContact")], axis=1)

# Calculate the 10 residues most frequently interacting with the ligand
top_10_residues = (
    df_filtered.T.groupby(level=["ligand", "protein"])
    .sum()
    .T.astype(bool)
    .mean()
    .sort_values(ascending=False)
    .head(10)
    .to_frame("%")
    .T
    * 100
)

top_10_residues
```

It worked well.

This next one however...

```
fp_count=plf.Fingerprint(count=True)
fp_count.run(u.trajectory[::10], ligand_selection, protein_selection)

fp_count.to_pickle("fingerprint.pkl")

# setup things to clean
to_cleanup = [
    ("UNL210", "SIA209", "VdWContact"),
]

# start cleanup
df = fp_count.to_dataframe()
for lig, residue, interaction in to_cleanup:
    # locate all frames with this combination
    frames = df.index[df.loc[:, (lig, residue, interaction)]]
    for frame in frames:
        # delete interaction from registry
        del fp_count.ifp[frame][(lig, residue)][interaction]
```
I didn't misspell something did I?

```
KeyError                                  Traceback (most recent call last)
Cell In[42], [line 18](vscode-notebook-cell:?execution_count=42&line=18)
     [15](vscode-notebook-cell:?execution_count=42&line=15) frames = df.index[df.loc[:, (lig, residue, interaction)]]
     [16](vscode-notebook-cell:?execution_count=42&line=16) for frame in frames:
     [17](vscode-notebook-cell:?execution_count=42&line=17)     # delete interaction from registry
---> [18](vscode-notebook-cell:?execution_count=42&line=18)     del fp_count.ifp[frame][(lig, residue)][interaction]

KeyError: 'VdWContact'
```

**Reply by cbouy** [reply]
My bad I didn't account for the count fingerprint, just use `df = fp_count.to_dataframe(count=False)` so that the temporary dataframe `df` used to locate frames with the specified interaction can be indexed correctly

**Reply by ONISniperOne** [reply]
So.... yeah this is also kinda related to this other issue #174 I brought up (which is solved). I have a ligand broken into 2 residues, but I'm only interested in the interactions of one part of the ligand with the rest of the protein. If I keep 2 resid's separate and put forth as one ligand, the whole thing is drawn correctly. If I only put the one part of the ligand I'm interested in, it's drawn incorrectly and affects SMARTS identification/fingerprint mapping too I think?

```
# Load topology and trajectory
u = mda.Universe('cocomplex_nowat_noion.prmtop', 'md.nc')

# Create selections for the ligand and protein
ligand_selection = u.select_atoms("resid 209:210")
protein_selection = u.select_atoms("resid 1:208", ligand=ligand_selection)
ligand_selection, protein_selection

fp_count=plf.Fingerprint(count=True)
fp_count.run(u.trajectory[::10], ligand_selection, protein_selection)

fp_count.to_pickle("fingerprint.pkl")

# setup things to clean
to_cleanup = [
    ("SIA209", "resid 1:208", ['Anionic',
 'CationPi',
 'Cationic',
 'EdgeToFace',
 'FaceToFace',
 'HBAcceptor',
 'HBDonor',
 'Hydrophobic',
 'MetalAcceptor',
 'MetalDonor',
 'PiCation',
 'PiStacking',
 'VdWContact',
 'XBAcceptor',
 'XBDonor']),
]

# start cleanup
df = fp_count.to_dataframe(count=False)
for lig, residue, interaction in to_cleanup:
    # locate all frames with this combination
    frames = df.index[df.loc[:, (lig, residue, interaction)]]
    for frame in frames:
        # delete interaction from registry
        del fp_count.ifp[frame][(lig, residue)][interaction]
```

I'm sorry I've been bothering you so much! I've been trying different combos.
To clarify, my protein is resid 1:208. Ligand is resid 209:210

Have been experimenting w & w/o intramolecular bonds and correctly drawing my ligand in part or whole.

**Reply by cbouy** [reply]
You're mixing things up for the `to_cleanup` part, the syntax should be a list of tuple containing 3 strings corresponding to what you can see in the dataframe, so the 2nd item `resid 1:208` won't work (that's the syntax for MDAnalysis selections) and should be something like `ALA123` (and you can't use a range here).

But since in your case you want to drop all interactions with a specific residue we can simplify the above cleanup code to just:
```python
for frame, ifp in fp_count.ifp.items():
    fp_count.ifp[frame] = ifp["SIA210"]  # only keep SIA210
```

However you'll still get the ligand with a missing bond between both residues in the 3D view since it only plots residues involved in an interaction, not the entire molecule. You can modify the ligand molecule object to include SIA209 inside SIA210 but it's quite convoluted:
```python
from rdkit import Chem

ligand_mol = plf.Molecule.from_mda(ligand_selection)
mol = ligand_mol["SIA210"]
combined = Chem.CombineMols(mol, ligand_mol["SIA209"])

with Chem.RWMol(combined) as rwmol:
# run once, then hover over both atoms with the missing bond to get the indices
# then modify the indices below, uncomment this line and re-run
#    rwmol.AddBond(0, 32, Chem.BondType.SINGLE)
    res = plf.residue.Residue(rwmol)
res.resid = mol.resid
ligand_mol.residues[plf.ResidueId.from_string("SIA210")] = res

fp.plot_3d(ligand_mol, protein_mol, frame=0)
```

**Reply by ONISniperOne** [reply]
![image](https://github.com/chemosim-lab/ProLIF/assets/128614019/0f8f1b36-fa4a-4cb4-ab9e-4b7f859a9398)

For some of my compounds, the entire ligand got is really far from the protein.

```
for frame, ifp in fp_count.ifp.items():
    fp_count.ifp[frame] = ifp["SIA210"]  # only keep SIA210
```
This part worked really well

**Reply by cbouy** [reply]
You have to rebuild both molecules for every frame that you want to plot e.g.
```python
frame = 0
# seek specific frame
u.trajectory[frame]
ligand_mol = plf.Molecule.from_mda(ligand_selection)
protein_mol = plf.Molecule.from_mda(protein_selection)
# display
view = fp_count.plot_3d(ligand_mol, protein_mol, frame=frame, display_all=False)
view
```
Otherwise it keeps the coordinates from the first frame of the trajectory
